protocols/virtualdefund/virtualdefund.go

```diff
@@ -145,6 +252,77 @@ func (o Objective) Update(event protocols.ObjectiveEvent) (protocols.Objective,
 	if o.Id() != event.ObjectiveId {
 		return &o, fmt.Errorf("event and objective Ids do not match: %s and %s respectively", string(event.ObjectiveId), string(o.Id()))
 	}
-	return &o, errors.New("TODO: UNIMPLEMENTED")
 
+	updated := o.clone()
+
+	for _, ss := range event.SignedStates {
+		incomingChannelId, _ := ss.State().ChannelId() // TODO handle error
+		vChannelId, _ := updated.VFixed.ChannelId()    // TODO handle error
+
+		if incomingChannelId != vChannelId {
+			return &o, errors.New("event channelId out of scope of objective")
+		} else {
+			incomingSignatures := ss.Signatures()
+			for i := uint(0); i < 3; i++ {
+				existingSig := o.Signatures[i]
+				incomingSig := incomingSignatures[i]
+
+				// If the incoming signature is zeroed we ignore it
+				if isZero(incomingSig) {
+					continue
+				}
+				// If the existing signature is not zeroed we check that it matches the incoming signature
+				if !isZero(existingSig) {
+					if existingSig.Equal(incomingSig) {
+						continue
+					} else {
+						return &o, fmt.Errorf("incoming signature %+v does not match existing %+v", incomingSig, existingSig)
+					}
+				}
+				// Otherwise we validate the incoming signature and update our signatures
+				isValid, err := updated.ValidateSignature(incomingSig, i)
+				if isValid {
+					// Update the signature
+					updated.Signatures[i] = incomingSig
+				} else {
+					return &o, fmt.Errorf("failed to validate signature: %w", err)
+				}
+			}
+		}
+	}
+	var toMyLeftId types.Destination
+	var toMyRightId types.Destination
+
+	if o.ToMyLeft != nil {
+		toMyLeftId = o.ToMyLeft.Id
+	}
+	if o.ToMyRight != nil {
+		toMyRightId = o.ToMyRight.Id
+	}
+
+	for _, sp := range event.SignedProposals {
+		var err error
+		switch sp.Proposal.ChannelID {
+		case types.Destination{}:
+			return &o, fmt.Errorf("signed proposal is for a zero-addressed ledger channel") // catch this case to avoid unspecified behaviour -- because if Alice or Bob we allow a null channel.
```

Comment:
```suggestion
			return &o, fmt.Errorf("signed proposal is for a zero-addressed ledger channel") // catch this case to avoid unspecified behaviour -- because of Alice or Bob we allow a null channel.
```